File: lib/py/telepact/internal/GetIncrementalApiDefinitions.py
#|
#|  Copyright The Telepact Authors
#|
#|  Licensed under the Apache License, Version 2.0 (the "License");
#|  you may not use this file except in compliance with the License.
#|  You may obtain a copy of the License at
#|
#|  https://www.apache.org/licenses/LICENSE-2.0
#|
#|  Unless required by applicable law or agreed to in writing, software
#|  distributed under the License is distributed on an "AS IS" BASIS,
#|  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#|  See the License for the specific language governing permissions and
#|  limitations under the License.
#|

import re
from typing import TYPE_CHECKING, cast

from .GetApiDefinitionsWithExamples import _get_schema_key

if TYPE_CHECKING:
    from ..TelepactSchema import TelepactSchema


_ENTRYPOINT_PREFIXES = ('info.', 'headers.', 'errors.', 'fn.')
_CUSTOM_TYPE_PREFIXES = ('fn.', 'struct.', 'union.', 'headers.', 'errors.', 'info.', '_ext.')


def get_api_entrypoint_definitions(
    telepact_schema: 'TelepactSchema',
    include_internal: bool,
) -> list[object]:
    definitions = telepact_schema.full if include_internal else telepact_schema.original
    return [
        definition
        for definition in definitions
        if _get_schema_key(cast(dict[str, object], definition)).startswith(_ENTRYPOINT_PREFIXES)
    ]


def get_api_definitions_by_schema_key(
    telepact_schema: 'TelepactSchema',
    schema_key: str,
    include_internal: bool,
) -> list[object] | None:
    definitions = telepact_schema.full if include_internal else telepact_schema.original
    definitions_by_key = {
        _get_schema_key(cast(dict[str, object], definition)): cast(dict[str, object], definition)
        for definition in definitions
    }
    if schema_key not in definitions_by_key:
        return None

    included_schema_keys: set[str] = set()
    pending_schema_keys: list[str] = [schema_key]

    while pending_schema_keys:
        current_schema_key = pending_schema_keys.pop()
        if current_schema_key in included_schema_keys:
            continue
        current_definition = definitions_by_key.get(current_schema_key)
        if current_definition is None:
            continue

        included_schema_keys.add(current_schema_key)
        pending_schema_keys.extend([
            referenced_schema_key
            for referenced_schema_key in _get_referenced_schema_keys(current_definition, definitions_by_key)
            if referenced_schema_key not in included_schema_keys
        ])

    return [
        definition
        for definition in definitions
        if _get_schema_key(cast(dict[str, object], definition)) in included_schema_keys
    ]


def _get_referenced_schema_keys(
    definition: dict[str, object],
    definitions_by_key: dict[str, dict[str, object]],
) -> list[str]:
    references: set[str] = set()

    for key, value in definition.items():
        if key in {'///', '_errors'}:
            continue
        _collect_referenced_schema_keys(value, definitions_by_key, references)

    errors_regex = definition.get('_errors')
    if isinstance(errors_regex, str):
        regex = re.compile(errors_regex)
        references.update([
            schema_key
            for schema_key in definitions_by_key.keys()
            if regex.match(schema_key)
        ])

    return sorted(references)


def _collect_referenced_schema_keys(
    value: object,
    definitions_by_key: dict[str, dict[str, object]],
    references: set[str],
) -> None:
    if isinstance(value, str):
        schema_key = value[:-1] if value.endswith('?') else value
        if schema_key.startswith(_CUSTOM_TYPE_PREFIXES) and schema_key in definitions_by_key:
            references.add(schema_key)
        return

    if isinstance(value, list):
        for entry in value:
            _collect_referenced_schema_keys(entry, definitions_by_key, references)
        return

    if isinstance(value, dict):
        for key, entry in value.items():
            if key == '///':
                continue
            _collect_referenced_schema_keys(entry, definitions_by_key, references)

File: lib/java/src/main/java/io/github/telepact/internal/GetIncrementalApiDefinitions.java
//|
//|  Copyright The Telepact Authors
//|
//|  Licensed under the Apache License, Version 2.0 (the "License");
//|  you may not use this file except in compliance with the License.
//|  You may obtain a copy of the License at
//|
//|  https://www.apache.org/licenses/LICENSE-2.0
//|
//|  Unless required by applicable law or agreed to in writing, software
//|  distributed under the License is distributed on an "AS IS" BASIS,
//|  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//|  See the License for the specific language governing permissions and
//|  limitations under the License.
//|

package io.github.telepact.internal;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;

import io.github.telepact.TelepactSchema;

public class GetIncrementalApiDefinitions {
    private static final List<String> ENTRYPOINT_PREFIXES = List.of("info.", "headers.", "errors.", "fn.");
    private static final List<String> CUSTOM_TYPE_PREFIXES = List.of("fn.", "struct.", "union.", "headers.", "errors.", "info.", "_ext.");

    public static List<Object> getApiEntrypointDefinitions(TelepactSchema telepactSchema, boolean includeInternal) {
        final var definitions = includeInternal ? telepactSchema.full : telepactSchema.original;
        return definitions.stream()
                .filter(definition -> hasPrefix(
                        GetApiDefinitionsWithExamples.getSchemaKey((Map<String, Object>) definition),
                        ENTRYPOINT_PREFIXES))
                .toList();
    }

    public static List<Object> getApiDefinitionsBySchemaKey(
            TelepactSchema telepactSchema,
            String schemaKey,
            boolean includeInternal) {
        final var definitions = includeInternal ? telepactSchema.full : telepactSchema.original;
        final var definitionsByKey = new LinkedHashMap<String, Map<String, Object>>();
        for (final var definition : definitions) {
            final var definitionMap = (Map<String, Object>) definition;
            definitionsByKey.put(GetApiDefinitionsWithExamples.getSchemaKey(definitionMap), definitionMap);
        }
        if (!definitionsByKey.containsKey(schemaKey)) {
            return null;
        }

        final var includedSchemaKeys = new LinkedHashSet<String>();
        final var pendingSchemaKeys = new ArrayList<String>(List.of(schemaKey));
        while (!pendingSchemaKeys.isEmpty()) {
            final var currentSchemaKey = pendingSchemaKeys.remove(pendingSchemaKeys.size() - 1);
            if (!includedSchemaKeys.add(currentSchemaKey)) {
                continue;
            }

            final var definitionMap = definitionsByKey.get(currentSchemaKey);
            if (definitionMap == null) {
                continue;
            }

            for (final var referencedSchemaKey : getReferencedSchemaKeys(definitionMap, definitionsByKey)) {
                if (!includedSchemaKeys.contains(referencedSchemaKey)) {
                    pendingSchemaKeys.add(referencedSchemaKey);
                }
            }
        }

        return definitions.stream()
                .filter(definition -> includedSchemaKeys.contains(
                        GetApiDefinitionsWithExamples.getSchemaKey((Map<String, Object>) definition)))
                .toList();
    }

    private static List<String> getReferencedSchemaKeys(
            Map<String, Object> definition,
            Map<String, Map<String, Object>> definitionsByKey) {
        final Set<String> references = new LinkedHashSet<>();
        for (final var entry : definition.entrySet()) {
            if (entry.getKey().equals("///") || entry.getKey().equals("_errors")) {
                continue;
            }
            collectReferencedSchemaKeys(entry.getValue(), definitionsByKey, references);
        }

        if (definition.get("_errors") instanceof String errorsRegex) {
            final var regex = Pattern.compile(errorsRegex);
            for (final var schemaKey : definitionsByKey.keySet()) {
                if (regex.matcher(schemaKey).matches()) {
                    references.add(schemaKey);
                }
            }
        }

        return references.stream().sorted().toList();
    }

    private static void collectReferencedSchemaKeys(
            Object value,
            Map<String, Map<String, Object>> definitionsByKey,
            Set<String> references) {
        if (value instanceof String stringValue) {
            final var schemaKey = stringValue.endsWith("?")
                    ? stringValue.substring(0, stringValue.length() - 1)
                    : stringValue;
            if (hasPrefix(schemaKey, CUSTOM_TYPE_PREFIXES) && definitionsByKey.containsKey(schemaKey)) {
                references.add(schemaKey);
            }
            return;
        }

        if (value instanceof List<?> listValue) {
            for (final var entry : listValue) {
                collectReferencedSchemaKeys(entry, definitionsByKey, references);
            }
            return;
        }

        if (value instanceof Map<?, ?> mapValue) {
            for (final var entry : mapValue.entrySet()) {
                if (String.valueOf(entry.getKey()).equals("///")) {
                    continue;
                }
                collectReferencedSchemaKeys(entry.getValue(), definitionsByKey, references);
            }
        }
    }

    private static boolean hasPrefix(String value, List<String> prefixes) {
        return prefixes.stream().anyMatch(value::startsWith);
    }
}
